### PR TITLE
Smaller bundle size

### PR DESCRIPTION
**Pull request by HonoluluHenk**
I started on better modularization/build size reduction, references:
* google-libphonenumber is huge #248
* libphonenumber.js is loaded without using the IsMobilePhone decorator #247
* Smaller build size #212

This means moving most decorators and their respective validation implementation to their own file.

Since this is very tedious work: please review and tell me if this is the way to go.

Nice side-effect: also adds an easy way to validate with just a function (see ValidateBy.ts).

### CODE REVIEW

File: test/functional/validation-functions-and-decorators.spec.ts [hunk 1]
```diff
@@ -2864,7 +2863,7 @@ describe("isPhoneNumber", function() {
             someProperty: string;
         }
 
-        it("should not fail if validator.validate said that its valid", function(done) {
+        it.only("should not fail if validator.validate said that its valid", function(done) {
```
Comment (henrikra) [inline test/functional/validation-functions-and-decorators.spec.ts]: Should you remove `only`?

**Comment by henrikra** [comment]
@HonoluluHenk Are you still alive? :D

**Comment by HonoluluHenk** [comment]
Sorry, currently on holidays :relaxed:

**Comment by NoNameProvided** [comment]
Hi @HonoluluHenk!

Yes, this is great work! I will help with the changes as my time allows. I have been inactive for a while now, but now I managed to spend time on OSS again (last week I have caught up with the changes in `class-transformer`, this week I will catch up with this lib. I will prioritize answers on issues first, but expect me to look/help with this on the weekend.)

**Comment by henrikra** [comment]
@NoNameProvided My pr would be easy to check out too :)

**Comment by henrikra** [comment]
@HonoluluHenk Can you fix conflicts?

**Comment by HonoluluHenk** [comment]
@NoNameProvided: migration done.
Most of the changed files are just the decorators and validation functions moved to their own file.
Files that require your special attention:
validation/ValidationExecutor.ts (I left one FIXME in validateOrReject)
validation/ValidationTypes.ts
validation/Validator.ts (validateValueByMetadata is gone since all metadata-validations are migrated to ValidateBy())
validation-schema/ValidationSchemaToMetadataTransformer.ts

**Comment by HonoluluHenk** [comment]
Measures with a minimal angular application:
Empty angular project without class-validator: 312k

Using class-validator 0.9.1: bundle total 864k

Compare to this refactoring:
Bundle size when adding class-validator: 328k (just using validate() and some simple type checkers)
Including @IsPhoneNumber: 772k (this is still less than v 0.9.1!)

**Comment by henrikra** [comment]
Wow I really like this. How do you measure bundle size btw?

**Comment by HonoluluHenk** [comment]
I think, the refactoring also (accidentally) fixes #287 and #292

**Comment by maxdumas** [comment]
I'd like to weigh in specifically regarding `google-libphonenumber` and `isPhoneNumber`. In addition to its size, that dependency is particularly problematic because it has a different license (a hybrid of MIT and Apache), and its `package.json` doesn't use a standard license name. This makes it a bit of a pain to work with in organizations that have automated license restriction policies that work off of the `package.json`. It would be great if that validator could be split off into its own package or otherwise made an optional dependency.

**Comment by JonDum** [comment]
This is such a large refactoring that this can easily block all other PRs while it looms (or it will get more and more difficult to keep in sync with other PRs). 

@NoNameProvided I'd implore you to consider merging this as is and letting community work out issues for a month or so before publishing it as a minor semver release. Maybe even publishing it under a `@next` tag for those that want to try it out.

**Comment by Djaler** [comment]
Bump

**Comment by Djaler** [comment]
@NoNameProvided @pleerock can you check this, please?

**Comment by henrikra** [comment]
VsCode is showing red when importing this big package 😬 
image

**Comment by quentinus95** [comment]
@NoNameProvided @pleerock when can we expect to get this merged and released?

**Comment by vlapo** [comment]
What about create `@next` branch and start working on more complex and api breaking features? It`s time to move.

**Comment by jbjhjm** [comment]
@NoNameProvided @pleerock Is class-validator still being developed actively?
I see last release has been almost a year ago, lots of unanswered issues and this important size issue is open for 9 months too. Don't get me wrong, I'm just worried if these things are still going to be addressed.

**Comment by quentinus95** [comment]
@jbjhjm most of typestack repositories won't be updated anymore. See https://github.com/typestack/routing-controllers/issues/461

**Comment by HorusGoul** [comment]
A few months ago, I forked this PR, fixed some issues and tested it against one of my apps that are currently running in production. Since then, I haven't faced any kind of issues so maybe some of you can use it as well.

Here's the repo: https://github.com/HorusGoul/class-validator
Builds get published in the [`latest`](https://github.com/HorusGoul/class-validator/tree/latest) branch. 

To install it with `npm` you can use the following command:

```bash
npm install HorusGoul/class-validator#latest
```

This is a workaround until this PR gets some attention from the project maintainers so I recommend you to fork my repo and maintain your own copy.

**Comment by henrikra** [comment]
@HonoluluHenk Can you fix one conflict?

**Comment by vlapo** [comment]
Hey folks, 

I love this PR. It is huge step forward. So @HonoluluHenk if you have still a little interest about PR I have a question.

I am wondering how another libraries handle bundle size? I think the possible answer is tree-shaking and correct bundling/packaging. Right now `class-validator` only compile `TS` into `JS` files. We should check another libraries and lib starters and try to improve our building process.

IMHO the best solution of this all is:
 - if tree-shaking and right bundling/packaging will resolve bundle size:
    - improve our build flow
    - merge changes in this PR - I think there are improvements in code structure and code quality
    - do not break public API and use old way of imports (from root folder) = no breaking change
 - if it will NOT resolve bundle size
    - review this PR
    - merge this PR in `@next` and mark as 1.0.0-alpha.0
    - start work on refactoring PRs and resolving issues

What do you all think? Kick me if I am wrong in some point. 
@HonoluluHenk, @HorusGoul do you have some time to check e.g. [rollup.js](https://rollupjs.org/guide/en/) and test my theory about tree-shaking and bundle size?


References:
 - https://github.com/Hotell/typescript-lib-starter#-features
 - https://github.com/alexjoverm/typescript-library-starter/blob/master/README.md#features
 - https://webpack.js.org/guides/tree-shaking
 - https://rollupjs.org/guide/en/

**Comment by lake2** [comment]
Does  this PR has been merged?  I find my bundle js contains libphonebumber.js but I don't import phone number.

**Comment by vlapo** [comment]
No it didnt. But this is my number one task for this repository. I will update our packaging and build process and allow tree-shaking. It should help. Stay tuned :)

**Comment by lake2** [comment]
> No it didnt. But this is my number one task for this repository. I will update our packaging and build process and allow tree-shaking. It should help. Stay tuned :)

Merge this PR as soon as possible will be appreciated. It has been almost one year since this PR has been opened. I saw the libphonenumber.ts take up a large proportion of the bundle.js in webpack bundle analyzer. It's a nigthmare of web project.

**Comment by maxime1992** [comment]
I've made a comment to show the impact on frontend here: https://github.com/typestack/class-validator/issues/248#issuecomment-521320117

And here's the result on a compressed/uglified frontend project with Angular to compare:

image

**Comment by Djaler** [comment]
Small workaround:
**webpack.config.js**
```
module.exports = {
  //...
  resolve: {
    alias: {
      'google-libphonenumber: path.resolve(__dirname, '../src/libphonenumber-stub.js'),
    }
  }
};
```
**libphonenumber-stub.js**
```
module.exports = {
    PhoneNumberUtil: {
        getInstance() {
            return null;
        },
    },
};
```

**Comment by ramyothman** [comment]
any updates here on when are you planning to merge this request?

**Comment by maxime1992** [comment]
For those using angular-cli and don't have access to webpack config as @Djaler shown above, I did a dirty hack :warning: in the meantime... :see_no_evil: 
 In my `package.json`, within the `scripts` property:

```
"postinstall": "echo \"\" > ./node_modules/google-libphonenumber/dist/libphonenumber.js || true"
```

\*flies away\*

**Comment by Djaler** [comment]
@maxime1992 in angular cli one can not change webpack config? In vue cli we can do it easily

**Comment by maxime1992** [comment]
> @maxime1992 in angular cli one can not change webpack config? In vue cli we can do it easily

You could eject or use ngx-build-plus for ex but it's definitely not worth it in this case as I do not use the library at all and a fix is coming. So I decided to go for the ugly quick fix there :)

**Comment by tmtron** [comment]
To get rid of libphonenumber for the angular-cli I came up with a hack. The idea is a combination of the comments from [maxime1992](https://github.com/typestack/class-validator/pull/258#issuecomment-521587401) and [Djaler](https://github.com/typestack/class-validator/pull/258#issuecomment-535068753)

In your workspace save this `libphonenumber-stub.js` file:

```js
module.exports = {
    PhoneNumberUtil: {
        getInstance() {
            return null;
        },
    },
};
```

in `package.json` use the `postinstall` hook to replace the original lib:

```json
{
  "scripts": {
    "postinstall": "cat libphonenumber-stub.js > ./node_modules/google-libphonenumber/dist/libphonenumber.js || true"
   }
}
```

**Comment by gempain** [comment]
@NoNameProvided any chance this PR would be merged in ?

**Comment by patrykkrawczyk** [comment]
please merge this @vlapo @NoNameProvided @quentinus95

**Comment by gempain** [comment]
@HonoluluHenk would you be able to rebase your changes on the latest target branch so the conflicts are resolved ?

**Comment by vlapo** [comment]
🚀 I would like to ask you all for help in testing and reviewing PR https://github.com/typestack/class-validator/pull/568 🚀

**Comment by vlapo** [comment]
@henrikra @gempain @patrykkrawczyk @maxime1992 @Djaler @ramyothman Could you please test preview release (https://github.com/typestack/class-validator/pull/568) and provide some feedback?

**Comment by gempain** [comment]
@vlapo yes, I'll give it a shot. Actually, had a chance yesterday to try 12.0.0.rc-3 with jest and I had a few issues, but it was not a clean install so I need to see if I can reproduce on a clean install. Give me a day or two so I can do this.

**Comment by patrykkrawczyk** [comment]
to me its still not tree shakeable, running this confirms it:
`npx agadoo dist/bundles/index.esm.js`

**Comment by gempain** [comment]
@vlapo I've put together a [demo repo](https://github.com/gempain/class-transformer-tree-shaking), and so far I can't get tree shaking to work with `0.12.0.refactor-4`. I followed webpack's [official tree shaking guide](https://webpack.js.org/guides/tree-shaking/) which tells you to set `"sideEffects": false"` in `package.json` and `optimization.usedExports` to `true` in `webpack.config.js`, but I'd be greatful if someone can take a look at the repo and see if everything seems configured properly.

@vlapo feel free to use my repo to test out your changes or fork it. Do you want me to add you as a contributor so you can commit there ?

**Comment by patrykkrawczyk** [comment]
it's also not tree-shakebale in my project
without class-validator: 4 kb
with class-validator: 903 kb
this is using 0.12.0.refactor-4 and only doing import of IsString decorator

**Comment by vlapo** [comment]
Thank you for testing! `0.12.0-refactor.5` is published.
@gempain I test it on your demo app. But I think you have to update tsconfig and add:
```
"module": "es2015",
"moduleResolution": "node",
```

@patrykkrawczyk could you test `0.12.0-refactor.5`? What is your environment? Do you use webpack in your project?

**Comment by gempain** [comment]
@vlapo after updating my tsconfig and upgrading to reafctor 5, it is now tree shakable ! Great job !

**Comment by patrykkrawczyk** [comment]
0.12.0-refactor.5 is tree shakeable in my project.
I'm writing AWS Lambda function using Serverless Framework, Node 12 and TS.
Without class-validator: 4.88KiB
With class-validator isString: 20.8 KiB
With class-validator @IsString: 21.6 KiB
With class-validator @IsPhoneNumber('DE'): 661 KB

**Comment by gempain** [comment]
@vlapo I think after this, we'll have to refactor the `@IsPhoneNumber` validator. I can't believe we're embedding 400kb just for validating a phone number. We'd need to drop `libphonenumber.js`.

**Comment by vlapo** [comment]
@gempain this PR is about keeping `libphonenumber.js`. After https://github.com/typestack/class-validator/pull/568 PR merged we can use any lib (even a large one) for validation as tree shaking will remove unused code.

**Comment by gempain** [comment]
@vlapo sure, now people that don't use the phone validator won't get those 400kb, but people that do use it will get those kb, and I find it ackward that the validation logic of the phone validator weighs as much as the entire library. Anyway, that's another matter.

**Comment by vlapo** [comment]
Of course. We have to inform users about this disadvantage (as a note in docs). Developers can use validatorjs `IsMobilePhone` validator as alternative. `IsPhoneNumber` is some kind of heavy gun 🔫 

But if someone know replacement for `libphonenumber.js` I am open for suggestions 🚀

**Comment by desmap** [comment]
Pls check my part of my bundle graph:

image

Is this normal? I've `import { IsMongoId, MaxLength, ArrayMaxSize } from 'class-validator` and `import { validate } from 'class-validator` in my codebase.

My tsconfig matches the one from @gempain and my webpack conf is the one from Next + `usedExports: true` and I added `"sideEffects": false"` in `package.json`.

FWIW, his is a Next production build of the client side. I have v0.12.2 of class-validator.

**Comment by ramyothman** [comment]
Class validator uses the phone validator lib from google which makes its
bundle big unfortunately

**Comment by jbjhjm** [comment]
@ramyothman should not be the case in most recent versions anymore. As you can see in @desmap 's chart, libphonenumber is not in there. I guess he/she is wondering why so many validators are bundled as just 3 were imported.

**Comment by NoNameProvided** [comment]
Just FYI, the package used for validating phone numbers has been replaced in the current develop branch with a package using the same meta information to validate the phone number but much smaller.

**Pull request closed without merging**